File: mobile/lib/connectors/dexscreener.dart
// Ported from packages/backend/src/connectors/dexscreener.ts.
import 'http_client.dart';

const _dexscreener = "https://api.dexscreener.com/latest/dex/tokens";

class DexScreenerPrice {
  final double price;
  final double liquidityUsd;
  final String venue;
  final int pairCount;
  DexScreenerPrice(this.price, this.liquidityUsd, this.venue, this.pairCount);
}

/// Independent second price source. Keyless, no rate-limit issues.
Future<DexScreenerPrice?> fetchDexPrice(String? contract) async {
  if (contract == null || !contract.startsWith('0x')) return null;

  final data = await fetchJson('$_dexscreener/$contract');
  final pairs = (data?['pairs'] as List?) ?? [];
  if (pairs.isEmpty) return null;

  final sorted = List<Map<String, dynamic>>.from(pairs)
    ..sort((a, b) {
      final la = num.tryParse('${(a['liquidity'] as Map?)?['usd'] ?? 0}') ?? 0;
      final lb = num.tryParse('${(b['liquidity'] as Map?)?['usd'] ?? 0}') ?? 0;
      return lb.compareTo(la);
    });
  final top = sorted.first;

  final price = double.tryParse('${top['priceUsd']}');
  if (price == null || !price.isFinite) return null;

  return DexScreenerPrice(
    price,
    num.tryParse('${(top['liquidity'] as Map?)?['usd'] ?? 0}')?.toDouble() ?? 0,
    '${top['dexId'] ?? 'unknown'} on ${top['chainId'] ?? 'unknown'}',
    sorted.length,
  );
}

File: packages/shared/src/connectors/dexscreener.ts
import { fetchJson } from "./base.js";

const DEXSCREENER = "https://api.dexscreener.com/latest/dex/tokens";

export interface DexScreenerPrice {
  price: number;
  liquidityUsd: number;
  venue: string;
  pairCount: number;
}

interface DexScreenerPair {
  priceUsd?: string;
  liquidity?: { usd?: number | string };
  dexId?: string;
  chainId?: string;
}

/** Independent second price source. Keyless, no rate-limit issues. */
export async function fetchDexPrice(
  contract: string | null | undefined
): Promise<DexScreenerPrice | null> {
  if (!contract || !contract.startsWith("0x")) return null;

  const data = await fetchJson<{ pairs?: DexScreenerPair[] }>(`${DEXSCREENER}/${contract}`);
  if (!data?.pairs?.length) return null;

  const pairs = [...data.pairs].sort(
    (a, b) => Number(b.liquidity?.usd ?? 0) - Number(a.liquidity?.usd ?? 0)
  );
  const top = pairs[0];

  const price = Number(top.priceUsd);
  if (!Number.isFinite(price)) return null;

  return {
    price,
    liquidityUsd: Number(top.liquidity?.usd ?? 0),
    venue: `${top.dexId ?? "unknown"} on ${top.chainId ?? "unknown"}`,
    pairCount: pairs.length,
  };
}
